### PR TITLE
Bug 1875939: vsphere - cloning a Windows 2019 (1909) template fails with Invalid operation for device '0'.

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
Add check for template disk size with tests.

Prevents user from hitting an undescriptive error message `invalid operation for device '0'`,
while attempting to create a machine from a template with a specified disk size larger
then the `machine.spec.diskGB` is requiring.

### CODE REVIEW

**Comment by enxebre** [comment]
thanks @Danil-Grigorev code changes looks good. Can you elaborate a bit on why the change is needed, why this gives an error from vSphere?
Can you please also elaborate the same raising and issue for `kubernetes-sigs/cluster-api-provider-vsphere` here https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/edd4fcfb973e0534b5f66edf4a851b543997e127/pkg/services/govmomi/vcenter/clone.go#L201

PTAL @jcpowermac

**Comment by Danil-Grigorev** [comment]
> Thanks @Danil-Grigorev code changes looks good. Can you elaborate a bit on why the change is needed, why this gives an error from vSphere?
> Can you please also elaborate the same raising and issue for `kubernetes-sigs/cluster-api-provider-vsphere` here https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/edd4fcfb973e0534b5f66edf4a851b543997e127/pkg/services/govmomi/vcenter/clone.go#L201
> 
> PTAL @jcpowermac

@enxebre I'll raise the issue upstream if there is a similar use-case when a user-specified disk capacity could be given to the clone method there. VSphere does not show an error of any kind anywhere, except the one quoted in the PR description, and the documentation says nothing about that issue. There is a guide on how to shrink disc down manually: https://virtualman.wordpress.com/2016/02/24/shrink-a-vmware-virtual-machine-disk-vmdk/ and a one to do this with specialized software: https://www.vmware.com/support/ws5/doc/ws_disk_shrink.html 

When we clone a VM template we are hitting the same constraints - no way to specify a lower disc size than the one already set on the template. So this PR implements a check, which evaluates whether you could set such size on the machine disk before calling an edit operation on its definition.

**Comment by Danil-Grigorev** [comment]
Raised the issue upstream: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/1030

File: pkg/controller/vsphere/reconciler.go [hunk 1]
```diff
@@ -616,7 +617,13 @@ func getDiskSpec(s *machineScope, devices object.VirtualDeviceList) (types.BaseV
 	}
 
 	disk := disks[0].(*types.VirtualDisk)
-	disk.CapacityInKB = int64(s.providerSpec.DiskGiB) * 1024 * 1024
+	cloneCapacityKB := int64(s.providerSpec.DiskGiB) * 1024 * 1024
+	if disk.CapacityInKB > cloneCapacityKB {
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/reconciler.go:621]: Should we not be requiring them to be the same? Does it being smaller not also cause an issue?

File: pkg/controller/vsphere/reconciler.go [hunk 2]
```diff
@@ -698,7 +705,7 @@ func taskIsFinished(task *mo.Task) (bool, error) {
 	case types.TaskInfoStateSuccess:
 		return true, nil
 	case types.TaskInfoStateError:
-		return true, nil
```
Comment (enxebre) [inline pkg/controller/vsphere/reconciler.go:701]: why this change?

File: pkg/controller/vsphere/reconciler.go [hunk 3]
```diff
@@ -616,7 +617,13 @@ func getDiskSpec(s *machineScope, devices object.VirtualDeviceList) (types.BaseV
 	}
 
 	disk := disks[0].(*types.VirtualDisk)
-	disk.CapacityInKB = int64(s.providerSpec.DiskGiB) * 1024 * 1024
+	cloneCapacityKB := int64(s.providerSpec.DiskGiB) * 1024 * 1024
+	if disk.CapacityInKB > cloneCapacityKB {
+		return nil, machineapierros.InvalidMachineConfiguration(
+			"can't resize template disk down, initial capacity is larger: %dKB > %dKB",
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/reconciler.go:623]: I think the values here are  in `KiB` as opposed to `KB`, we should try to make sure the error message has the correct units

File: pkg/controller/vsphere/reconciler.go (hunk 2 above)
Comment (Danil-Grigorev) [reply pkg/controller/vsphere/reconciler.go:701]: We were not reporting an error before that, so even getting this `invalid operation for device '0'` was possible only from vmware ui. I did this as otherwise we lose similar errors.

File: pkg/controller/vsphere/reconciler_test.go [hunk 4]
```diff
@@ -161,6 +163,23 @@ func TestClone(t *testing.T) {
...
+		{
+			testCase: "fail on disc resize down",
+			providerSpec: vsphereapi.VSphereMachineProviderSpec{
+				CredentialsSecret: &corev1.LocalObjectReference{
+					Name: "test",
+				},
+				Workspace: &vsphereapi.Workspace{
+					Server: server.URL.Host,
+				},
+				DiskGiB:  0,
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/reconciler_test.go:175]: Can we make this non-zero or add a separate case? I would expect a zero value to be caught earlier or maybe even produce a different error. While it may not now, we may change this later and that would affect that validity of this test.

File: pkg/controller/vsphere/reconciler.go (hunk 2 above)
Comment (Danil-Grigorev) [reply pkg/controller/vsphere/reconciler.go:701]: Oh, and the following `update` call on the machine fails with the inability to find VM by UUID, because the error here causes the VM to be removed automatically. This results in a nil panic in the code, and the controller goes into a crash loop.

File: pkg/controller/vsphere/reconciler.go (hunk 1 above)
Comment (Danil-Grigorev) [reply pkg/controller/vsphere/reconciler.go:621]: No, expanding the disk is a supported operation.

File: pkg/controller/vsphere/reconciler.go (hunk 3 above)
Comment (Danil-Grigorev) [reply pkg/controller/vsphere/reconciler.go:623]: Good catch, will fix that, but keep in mind it is an upstream issue for storage size pedantic people 
image
This is what vmware UI shows on the template. We should actually multiply by `1000` and call it `GB/MB/KB` etc.

File: pkg/controller/vsphere/reconciler.go (hunk 2 above)
Comment (enxebre) [reply pkg/controller/vsphere/reconciler.go:701]: thanks, please move to a different commit.

File: pkg/controller/vsphere/reconciler_test.go (hunk 4 above)
Comment (Danil-Grigorev) [reply pkg/controller/vsphere/reconciler_test.go]: Done.

File: pkg/controller/vsphere/reconciler.go (hunk 1 above)
Comment (JoelSpeed) [reply pkg/controller/vsphere/reconciler.go:621]: I was looking at this backwards, ignore me

**Comment by jcpowermac** [comment]
Just tested with @Danil-Grigorev cluster in VMC. I can confirm that Windows does clone and started up as expected.

File: pkg/controller/vsphere/reconciler.go (hunk 2 above)
Comment (elmiko) [reply pkg/controller/vsphere/reconciler.go:701]: i'm having trouble following the conversation here, should this be moved to another PR or not?

**Review by elmiko — COMMENTED** [review]
this makes sense to me, i just have a question

File: pkg/controller/vsphere/reconciler.go (hunk 2 above)
Comment (Danil-Grigorev) [reply pkg/controller/vsphere/reconciler.go:701]: @elmiko The change is already in its own commit, so this PR is just a holder for both atm. Also Joel hit this issue, and we discussed that the fix will be contained in this PR https://coreos.slack.com/archives/GE2HQ9QP4/p1599829591211300?thread_ts=1599829162.210500&channel=GE2HQ9QP4&message_ts=1599829591.211300
